File: src/test/groovy/graphql/NestedInputSchema.java
package graphql;

import graphql.schema.DataFetcher;
import graphql.schema.DataFetchingEnvironment;
import graphql.schema.GraphQLArgument;
import graphql.schema.GraphQLFieldDefinition;
import graphql.schema.GraphQLInputObjectField;
import graphql.schema.GraphQLInputObjectType;
import graphql.schema.GraphQLObjectType;
import graphql.schema.GraphQLSchema;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

import static graphql.Scalars.GraphQLBoolean;
import static graphql.Scalars.GraphQLInt;

public class NestedInputSchema {


    public static GraphQLSchema createSchema() {


        GraphQLObjectType root = rootType();

        return GraphQLSchema.Companion.newSchema()
                                      .query(root)
                                      .build();
    }

    public static GraphQLObjectType rootType() {
        return GraphQLObjectType.Companion.newObject()

                                          .name("Root")
                                          .field(GraphQLFieldDefinition.Companion.newFieldDefinition()
                                                                                 .name("value")
                                                                                 .type(INSTANCE.getGraphQLInt())
                                                                                 .dataFetcher(new DataFetcher() {
                            @Override
                            public CompletionStage<Object> fetch(DataFetchingEnvironment environment) {
                                Integer initialValue = environment.argument("initialValue");
                                Map<String, Object> filter = environment.argument("filter");
                                if (filter != null) {
                                    if (filter.containsKey("even")) {
                                        Boolean even = (Boolean) filter.get("even");
                                        if (even && (initialValue%2 != 0)) {
                                            return CompletableFuture.completedFuture(0);
                                        } else if (!even && (initialValue%2 == 0)) {
                                            return CompletableFuture.completedFuture(0);
                                        }
                                    }
                                    if (filter.containsKey("range")) {
                                        Map<String, Integer> range = (Map<String, Integer>) filter.get("range");
                                        if (initialValue < range.get("lowerBound") ||
                                                initialValue > range.get("upperBound")) {
                                            return CompletableFuture.completedFuture(0);
                                        }
                                    }
                                }
                                return CompletableFuture.completedFuture(initialValue);
                            }})
                                                                                 .argument(GraphQLArgument.Companion.newArgument()
                                                           .name("intialValue")
                                                           .type(INSTANCE.getGraphQLInt())
                                                           .defaultValue(5))
                                                                                 .argument(GraphQLArgument.Companion.newArgument()
                                                           .name("filter")
                                                           .type(filterType())))
                                          .build();
    }

    public static GraphQLInputObjectType filterType() {
        return GraphQLInputObjectType.Companion.newInputObject()
                                               .name("Filter")
                                               .field(GraphQLInputObjectField.Companion.newInputObjectField()
                                                        .name("even")
                                                        .type(INSTANCE.getGraphQLBoolean()))
                                               .field(GraphQLInputObjectField.Companion.newInputObjectField()
                                                        .name("range")
                                                        .type(rangeType()))
                                               .build();
    }

    public static GraphQLInputObjectType rangeType() {
        return GraphQLInputObjectType.Companion.newInputObject()
                                               .name("Range")
                                               .field(GraphQLInputObjectField.Companion.newInputObjectField()
                                                        .name("lowerBound")
                                                        .type(INSTANCE.getGraphQLInt()))
                                               .field(GraphQLInputObjectField.Companion.newInputObjectField()
                                                        .name("upperBound")
                                                        .type(INSTANCE.getGraphQLInt()))
                                               .build();
    }
}

File: src/test/groovy/graphql/NestedInputSchema.kt
package graphql

import graphql.schema.newInputObject
import graphql.schema.newObject
import graphql.schema.newSchema
import java.util.concurrent.CompletableFuture.completedFuture


val rangeType = newInputObject {
    name = "Range"
    field {
        name = "lowerBound"
        type = GraphQLInt
    }
    field {
        name = "upperBound"
        type = GraphQLInt
    }
}

val filterType = newInputObject {
    name = "Filter"
    field {
        name = "even"
        type = GraphQLBoolean
    }
    field {
        name = "range"
        type = rangeType
    }
}

val rootType = newObject {
    name = "Root"
    field<Int> {
        name = "value"
        argument {
            name = "initialValue"
            type = GraphQLInt
            defaultValue = 5
        }
        argument {
            name = "filter"
            type = filterType
        }
        fetcher = { environment ->
            val initialValue = environment.argument<Int>("initialValue")!!
            val filter = environment.argument<Map<String, Any>>("filter")!!

            if (filter.containsKey("even")) {
                val even = filter["even"] as Boolean
                if (even && initialValue % 2 != 0) {
                    completedFuture<Int>(0)
                } else if (!even && initialValue % 2 == 0) {
                    completedFuture<Int>(0)
                }
            } else if (filter.containsKey("range")) {
                val range = filter["range"] as Map<String, Int>
                if (initialValue < range["lowerBound"]!! || initialValue > range["upperBound"]!!) {
                    completedFuture<Int>(0)
                }
            }

            completedFuture<Int>(initialValue)
        }
    }
}
val nestedSchema = newSchema {
    query = rootType
}